### PR TITLE
Added Hyper-V Android Emulator Setup Instructions. Closes #3234

### LINKED ISSUE #3234: [Windows] Add Hyper-V Android Emulator to react-native documentation.
A lot of developers, especially in the enterprise space are using Hyper-V on their machines. Unfortunately the default Android Emulator cannot be hardware accelerated if Hyper-V is turned on and turning it off is often not an option.

Fortunately there is a great and free android emulator that uses Hyper-V.

The [Visual Studio Emulator for Android](https://www.visualstudio.com/en-us/features/msft-android-emulator-vs.aspx). The [standalone download link](http://aka.ms/vscomemudownload) is at the bottom of that page. It doesn't require you to install Visual Studio at all.

It is free, very powerful, fast and you don't have to turn off your Hyper-V to use the default Android Emulator.

I think this information and the following instruction should make it into the official `react-native` `getting started guide`.

To use it with react-native you just have to add a key and value to your registry:

HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Android SDK Tools

String Value:
Name: Path
Value: C:\Program Files\Android\sdk

Obviously the path value might be different on your machine.

This emulator also need the `adb reverse tcp:8081 tcp:8081` command executed once before starting.

Then restart the emulator and when it is running you can just do `react-native run-android` as usual.

visual studio emulator regedit

### PR DESCRIPTION
**Pull request by BerndWessels**
These are the instructions to use a Hyper-V hardware accelerated Android emulator on Windows.

### CODE REVIEW

File: docs/DevelopmentSetupAndroid.md [hunk 1]
```diff
@@ -78,3 +78,29 @@ Genymotion is much easier to set up than stock Google emulators. However, it's o
...
+
+Fortunately there is a great and free android emulator that uses Hyper-V.
+
+The [Visual Studio Emulator for Android](https://www.visualstudio.com/en-us/features/msft-android-emulator-vs.aspx). The [standalone download link](http://aka.ms/vscomemudownload) is at the bottom of that page. It doesn't require you to install Visual Studio at all.
+
+It is free, very powerful, fast and hardware accelerated via Hyper-V technology.
+
+To use it with react-native you just have to add a key and value to your registry:
+
+HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Android SDK Tools
```
Comment (satya164) [inline docs/DevelopmentSetupAndroid.md]: Can you wrap it in backticks? like

    ```
    HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Android SDK Tools
    ```

File: docs/DevelopmentSetupAndroid.md [hunk 2]
```diff
@@ -78,3 +78,29 @@ Genymotion is much easier to set up than stock Google emulators. However, it's o
   ![Create AVD dialog](/react-native/img/CreateAVD.png)
   2. With the new AVD selected, click `Start...`
 5. To bring up the developer menu press F2
+
+### Windows Hyper-V Alternative
+
+A lot of developers, especially in the enterprise space are using Hyper-V on their machines. Unfortunately the default Android Emulator cannot be hardware accelerated if Hyper-V is turned on and turning it off is often not an option.
+
+Fortunately there is a great and free android emulator that uses Hyper-V.
```
Comment (satya164) [inline docs/DevelopmentSetupAndroid.md]: The [Visual Studio Emulator for Android](https://www.visualstudio.com/en-us/features/msft-android-emulator-vs.aspx) is a free android emulator that is hardware accelerated via Hyper-V. It doesn't require you to install Visual Studio at all.

File: docs/DevelopmentSetupAndroid.md [hunk 3]
```diff
@@ -78,3 +78,29 @@ Genymotion is much easier to set up than stock Google emulators. However, it's o
...
+It is free, very powerful, fast and hardware accelerated via Hyper-V technology.
+
+To use it with react-native you just have to add a key and value to your registry:
+
+HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Android SDK Tools
+
+String Value:
+Name: Path
+Value: C:\Program Files\Android\sdk
+
```
Comment (satya164) [inline docs/DevelopmentSetupAndroid.md]: Also this,

    ```
    String Value:
    Name: Path
    Value: C:\Program Files\Android\sdk
    ```

It'll also be helpful to add the exact steps to do this. For example, Open run command, type rgedit.exe etc.

File: docs/DevelopmentSetupAndroid.md [hunk 4]
```diff
@@ -78,3 +78,29 @@ Genymotion is much easier to set up than stock Google emulators. However, it's o
...
+Obviously the path value might be different on your machine.
+
+This emulator also need the `adb reverse tcp:8081 tcp:8081` command executed once before starting.
```
Comment (satya164) [inline docs/DevelopmentSetupAndroid.md]:
```
You will also need to run the command `adb reverse tcp:8081 tcp:8081` with this emulator.
```

File: docs/DevelopmentSetupAndroid.md [hunk 5]
```diff
@@ -78,3 +78,29 @@ Genymotion is much easier to set up than stock Google emulators. However, it's o
...
+
+To use it with react-native you just have to add a key and value to your registry:
+
+HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Android SDK Tools
+
+String Value:
+Name: Path
+Value: C:\Program Files\Android\sdk
+
+Obviously the path value might be different on your machine.
```
Comment (satya164) [inline docs/DevelopmentSetupAndroid.md]: Remove 'Obviously'.

**Comment by satya164** [comment]
Thanks for the PR. Let's fix these few thinsg before merging.

@BerndWessels What about the assets? Where should we upload them?

**Comment by BerndWessels** [comment]
@satya164 I've updated based on your feedback. Not sure if the screenshot is actually needed.

**Comment by astreet** [comment]
Thanks! Assets should be uploaded to {RN_REPO_ROOT}/website/src/react-native/img

File: docs/DevelopmentSetupAndroid.md [hunk 6]
```diff
@@ -78,3 +78,28 @@ Genymotion is much easier to set up than stock Google emulators. However, it's o
   ![Create AVD dialog](/react-native/img/CreateAVD.png)
   2. With the new AVD selected, click `Start...`
 5. To bring up the developer menu press F2
+
+### Windows Hyper-V Alternative
+
+A lot of developers, especially in the enterprise space are using Hyper-V on their machines. Unfortunately the default Android Emulator cannot be hardware accelerated if Hyper-V is turned on and turning it off is often not an option.
+
+Fortunately there is a great and free android emulator that uses Hyper-V.
```
Comment (satya164) [inline docs/DevelopmentSetupAndroid.md]: Can you remove this line? We already say these things in the paragraph below.

**Comment by satya164** [comment]
Thanks @astreet 

@BerndWessels Can you add the image to the location @astreet suggested before we can merge ?

**Comment by BerndWessels** [comment]
@satya164 I removed the lines you mentioned and also the screenshot since the documentation now explains all the steps in detail.

**Comment by satya164** [comment]
@BerndWessels Thanks. Looks good. Can you squash the commits to one before we merge?

File: docs/DevelopmentSetupAndroid.md [hunk 7]
```diff
@@ -78,3 +78,22 @@ Genymotion is much easier to set up than stock Google emulators. However, it's o
   ![Create AVD dialog](/react-native/img/CreateAVD.png)
   2. With the new AVD selected, click `Start...`
 5. To bring up the developer menu press F2
+
+### Windows Hyper-V Alternative
+
+The [Visual Studio Emulator for Android](https://www.visualstudio.com/en-us/features/msft-android-emulator-vs.aspx) is a free android emulator that is hardware accelerated via Hyper-V. It doesn't require you to install Visual Studio at all.
+
+It is free, very powerful, fast and hardware accelerated via Hyper-V technology.
```
Comment (satya164) [inline docs/DevelopmentSetupAndroid.md]: Okay, missed this one also. We basically say the same things in the previous paragraph. So remove this also.

**Comment by BerndWessels** [comment]
@satya164 oops, how do I do that?

**Comment by satya164** [comment]
@BerndWessels http://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git

**Comment by BerndWessels** [comment]
@satya164 I hope the squashing worked, the text should be fine now and the screenshot is removed.

**Comment by satya164** [comment]
@BerndWessels Nope. There are still 5 commits.

**Comment by satya164** [comment]
Also, you didn't remove the line I mentioned. :(

**Comment by BerndWessels** [comment]
@satya164 OK, so this should be the text, right? Or which line do you mean?

### Windows Hyper-V Alternative

The [Visual Studio Emulator for Android](https://www.visualstudio.com/en-us/features/msft-android-emulator-vs.aspx) is a free android emulator that is hardware accelerated via Hyper-V. It doesn't require you to install Visual Studio at all.

To use it with react-native you just have to add a key and value to your registry:

1. Open the Run Command (Windows+R)
2. Enter `regedit.exe`
3. In the Registry Editor navigate to `HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Android SDK Tools`
4. Right Click on `Android SDK Tools` and choose `New > String Value`
5. Set the name to `Path`
6. Double Click the new `Path` Key and set the value to `C:\Program Files\Android\sdk`. The path value might be different on your machine.

You will also need to run the command `adb reverse tcp:8081 tcp:8081` with this emulator.

Then restart the emulator and when it runs you can just do `react-native run-android` as usual.

**Comment by satya164** [comment]
Yup.

**Comment by satya164** [comment]
I meant this line - https://github.com/facebook/react-native/pull/5117/files#diff-361847dc7e4835da109d8fc6eac71558R86

**Comment by BerndWessels** [comment]
@satya164 Sorry I doubt I get the commits squashed together since I made them in the master of my fork rather than in a branch - at least I can't find anyway to achieve that. If all this is too much trouble I can just create a new pull request or you make your own based on whats here. Sorry for the trouble - I sometimes struggle with Git.

**Comment by satya164** [comment]
@BerndWessels No problem. Branch shouldn't make a differenc.

You can alternatively run the following,

```sh
git reset --soft HEAD~6
git commit -m "Added Hyper-V Android Emulator Setup Instructions. Closes #3234"
git push -f
```

If it doesn't work. Feel free to send another PR and tag me.

**Comment by BerndWessels** [comment]
@satya164 ok, I did the git commands, now there is only one commit but I think there is other stuff in it now that does not belong to it.

**Comment by satya164** [comment]
Ouch. Just do a new PR. Fixing is too much work. Sorry for the trouble.

**Comment by mkonicek** [comment]
Thanks for adding this!

**Comment by satya164** [comment]
@BerndWessels Any updates on this?

**Comment by BerndWessels** [comment]
@satya164 Hi, here is [a new PR](https://github.com/facebook/react-native/pull/5239) for this - we should close and abandon this one I guess.

**Pull request closed without merging**